man/man1/mtutil.1
.TH MTUTIL 1 "March 2021" "Version 2021.03.01" "User Manuals"

.SH NAME
mtutil \- manipulate earthquake moment tensor, magnitude, and kinematic parameters

.SH SYNOPSIS
.P
.B mtutil -sdr|-mij|-pnt|-mag|-mom
.BI [ input ]
.B -sdr|-mij|-pnt|-mag|-mom|-ternary|-dcp|-sv|-kagan
.BI [ output ]
.BI [...]

.SH DESCRIPTION
.B mtutil
converts between earthquake source parameters. Inputs can be strike/dip/rake,
moment tensor components, moment tensor principal axes, seismic moment, or magnitude. Outputs can be
any of these values (when possible, e.g., you cannot convert a seismic moment to strike/dip/rake), a
ternary representation of the focal mechanism (Frohlich, 1992), double-couple percentage, slip vectors,
or the angle between two sources (Tape and Tape, 2012).
See Jost and Herrmann (1989) for more details.

If
.I input
is a file name, then
.B mtutil
will attempt to read inputs from that file.
If
.I input
is a comma-separated list of numbers, then inputs will be read this list of arguments.
If
.I input
is not provided, then
.B mtutil
will read from standard input.
.B mtutil
will print the results to standard output unless a file name is defined as
.IR output .


.SH OPTIONS
.TP
.B INPUTS

.TP
.BI "-sdr [" input_file | strike , dip , rake ]
.IR Strike ", " dip ", and " rake
as defined by Aki and Richards (2002).
.I Strike
is in degrees clockwise from north,
.I dip
is in degrees down from horizontal, and
.I rake
is in degrees counter-clockwise from the strike-parallel direction along the fault plane.

.TP
.BI "-mij [" input_file | mrr , mtt , mpp , mrt , mrp , mtp ]
Moment tensor components as published by the Global CMT project
(Gilbert and Dziewonski, 1975; Ekstrom et al., 2012), where
.I r
is the radial (unit vector corresponds to vertical) direction,
.I t
is co-latitude (unit vector corresponds to south), and
.I p
is longitude (unit vector corresponds to east).

Ekstrom et al. (2012) define a geographic Cartesian coordinate system for the
Global Centroid Moment Tensor project,
where
.IR x =north,
.IR y =east,
and
.IR z =down.
Some moment tensor solutions are reported in this
system (e.g., NIED F-net solutions from Japan). In this case:
.br
.IR mrr " =  mzz"
.br
.IR mtt " =  mxx"
.br
.IR mpp " =  myy"
.br
.IR mrt " =  mxz"
.br
.IR mrp " = -myz"
.br
.IR mtp " = -mxy"

Moment tensors behave similarly to stress tensors. You can use
.B mtutil
to compute the eigenvectors and eigenvalues of stress tensors using -mij and -pnt,
remembering to account for the coordinate transformation.
In local
.IR e ast,
.IR n orth,
and
.IR v ertical
(positive up) coordinates, the components of the tensor become:
.br
.IR mrr " =  mvv"
.br
.IR mtt " =  mnn"
.br
.IR mpp " =  mee"
.br
.IR mrt " = -mnv"
.br
.IR mrp " =  mev"
.br
.IR mtp " = -men"
.br
The eigenvectors computed by
.B -pnt
are reported in the same x=east, y=north, z=vertical reference frame.

.TP
.BI "-pnt [" input_file | px , py , pz , nx , ny , nz , tx , ty , tz , pmag , nmag , tmag ]
Moment tensor eigenvectors and eigenvalues, also referred to as P (maximum compression), N
(neutral), and T (maximum tension) axes. Note:
.I pmag
is typically negative and
.I tmag
is typically positive.

.TP
.BI "-mag [" input_file | magnitude ]
Moment
.IR magnitude " = 2/3*log10(" moment ")-6.0,"
where the seismic moment is in N*m.

.TP
.BI "-mom [" input_file | moment ]
Seismic
.I moment
in N*m.

.TP
.B OUTPUTS
.TP
.BI "-sdr [" output_file ]
When used as the output option with
.B -sdr
as the input, prints
.IR strike ", " dip ", and " rake
of the second nodal plane. When used as the output option with
.B -mij
or
.BR -pnt ,
prints results for both nodal planes.
.TP
.BI "-mij [" output_file ]
Moment tensor components:
.IR  mrr ", " mtt ", " mpp ", " mrt ", " mrp ", and " mtp
.TP
.BI "-pnt [" output_file ]
Moment tensor eigenvectors (normalized to length=1) and eigenvalues:
.IR  px ", " py ", " pz ", " nx ", " ny ", " nz ", " tx ", " ty ", " tz ", " pmag ", " nmag ", and " tmag
.TP
.BI "-mag [" output_file ]
Moment
.IR magnitude " = 2/3*log10(" moment ")-6.0"
.TP
.BI "-mom [" output_file ]
Seismic
.I moment
in N*m.
.TP
.BI "-ternary [" output_file ]
Earthquake mechanism characterization from Frohlich (1992):
.IR fraction_thrust ", " fraction_strike-slip ", and " fraction_normal ,
which add up to one.
.TP
.BI "-dcp [" output_file ]
Double couple percentage:
.IR dcp =min(1-2*abs( nmag / pmag ),1-2*abs( nmag / tmag ))
.TP
.BI "-sv [" output_file ]
Slip vector
.IR east ,
.IR north ,
and
.I vertical
components
.TP
.BI "-kagan [" output_file ]
Angle between two sources, after Tape and Tape (2012). This is the minimum angle required to rotate
the first source into the second source. The input must contain two sources instead of one.

.TP
.B OTHER OPTIONS
.TP
.B -[non-]dc
The default behavior of
.B mtutil
is
.BR -dc ,
to use the scalar seismic moment for a double couple source to calculate the seismic moment
and moment magnitude, as defined in Jost and Herrmann (1989) Equation 19:
.br
    M0 = 1/2 * (|m1| + |m2|)
.br
where m1 and m2 are the two largest eigenvalues of the moment tensor (in an absolute sense). This
is the definition of the seismic moment used by the Global Centroid Moment Tensor catalog.
Specify
.B -non-dc
to instead use the scalar seismic moment for a full moment tensor (Jost and Herrmann, 1989
Equation 20):
.br
    M0 = sqrt((m1^2 + m2^2 + m3^2)/2)
.br
where m1, m2, and m3 are the eigenvalues of the moment tensor. This is the definition used by the
USGS and the scaling factor for GMT psmeca.

.SH REFERENCES
.TP
Aki, K., and Richards, P.G. (2002).
Quantitative Seismology. University Science Books.
.TP
Frohlich, C. (1992).
Triangle diagrams: ternary graphs to display similarity and diversity of
earthquake focal mechanisms. Physics of the Earth and Planetary Interiors 75, 193-198.
.TP
Gilbert, F., and Dziewonski, A.M. (1975).
An application of normal mode theory to the retrieval of structural parameters and
source mechanisms from seismic spectra.
Philosophical Transactions of the Royal Society of London. Series A, 278, 187–269.
.TP
Jost, M.L., and Herrmann, R.B. (1989).
A student's guide to and review of moment tensors.
Seismological Research Letters 60, 37-57.
.TP
Tape, W., and Tape, C. (2012).
Angle between principal axis triples.
Geophysical Journal International 191, 813-831.

.RS


src/mtutil.f90
module mtutil

character(len=32) :: mtutil_mode
character(len=512) :: input
character(len=32) :: input_mode
character(len=512) :: output
integer :: nvals_input
integer :: nvals_output
character(len=5) :: scalar_moment_mode

end module

!==================================================================================================!

program main
!----
! Program for manipulating seismic source data, including:
!     Moment tensors
!     P, N, and T axes
!     Double couple parameters
!     Seismic moment/Moment magnitude
!     Ternary characterization (after Frohlich, 1992)
!----

use io, only: stdin, stdout, stderr, fileExists, line_count
use eq, only: mag2mom, &
              mom2mag, &
              mij2dcp, &
              mij2mag, &
              mij2mom, &
              mij2pnt, &
              mij2sdr, &
              mij2sv, &
              mij2ter, &
              pnt2dcp, &
              pnt2mag, &
              pnt2mag_nondc, &
              pnt2mij, &
              pnt2mom, &
              pnt2mom_nondc, &
              pnt2sdr, &
              pnt2sv, &
              pnt2ter, &
              sdr2mij, &
              sdr2pnt, &
              sdr2sdr, &
              sdr2sv, &
              sdr2ter, &
              sdr2kagan, &
              mij2kagan

use mtutil, only: mtutil_mode, &
                  input, &
                  input_mode, &
                  output, &
                  nvals_input, &
                  nvals_output, &
                  scalar_moment_mode
implicit none

! Local variables
integer :: i, j, ninputs, luin, luout, ios
double precision :: input_values(12), input_values2(12), output_values(12)
character(len=512) :: tmp_file, input_line
character(len=64) :: fmt_string
character(len=8) :: output_mode


! Initialize variables
input_values = 0.0d0
output_values = 0.0d0
tmp_file = 'mtutil_86_this_file_when_program_exits.tmp'


! Parse the command line
call gcmdln()


! Count the number of input lines and set the input unit
if (input_mode.eq.'file') then
    luin = 11
    if (.not.fileExists(input)) then
        call usage('mtutil: input file "'//trim(input)//'" not found')
    endif
    ninputs = line_count(input)
elseif (input_mode.eq.'cmdln') then
    luin = 11
    ! Replace commas with spaces in command line argument
    i = index(input,',')
    do while (i.ne.0)
        input(i:i) = ' '
        i = index(input,',')
    enddo
    ! Write command line info to temporary input file
    open(unit=luin,file=tmp_file,status='unknown')
    write(luin,*) input
    close(luin)
    input = tmp_file
    ninputs = 1
elseif (input_mode.eq.'stdin') then
    luin = stdin
    ninputs = 100000000
else
    call usage('mtutil: no input_mode named "'//trim(input_mode)//'"')
endif


! Open the input file if necessary
if (input_mode.eq.'file'.or.input_mode.eq.'cmdln') then
    open(unit=luin,file=input,status='old')
endif

! Set the output unit
if (output.eq.'stdout') then
    luout = stdout
else
    luout = 12
    open(unit=luout,file=output,status='unknown')
endif


! Read the inputs and run specified calculation
do i = 1,ninputs
    read(luin,'(A)',end=1001) input_line
    if (nvals_input.ge.1) then
        read(input_line,*,iostat=ios) (input_values(j),j=1,nvals_input)
    elseif (nvals_input.le.-1) then
        read(input_line,*,iostat=ios) (input_values(j),j=1,abs(nvals_input)), &
                                      (input_values2(j),j=1,abs(nvals_input))
    else
        write(stderr,*) 'nvals_input set to illegal value'
    endif
    if (ios.ne.0) then
        write(stderr,*) 'mtutil: error parsing inputs "',trim(input_line),'"'
        call usage('')
    endif
    ! write(0,*) input_values(1:nvals_input)

    ! Convert between moment and magnitude
    if (mtutil_mode.eq.'mag2mom') then
        call mag2mom(input_values(1),output_values(1))
    elseif (mtutil_mode.eq.'mom2mag') then
        call mom2mag(input_values(1),output_values(1))

    ! Input quantity: moment tensor
    elseif (mtutil_mode.eq.'mij2dcp') then
        call mij2dcp(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values(1))
    elseif (mtutil_mode.eq.'mij2mag') then
        call mij2mag(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values(1),scalar_moment_mode)
        if (abs(output_values(1)).le.-99.0d0) then
            call usage('mtutil: no scalar moment mode called '//trim(scalar_moment_mode))
        endif
    elseif (mtutil_mode.eq.'mij2mom') then
        call mij2mom(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values(1),scalar_moment_mode)
        if (abs(output_values(1)).le.1.0d-50) then
            call usage('mtutil: no scalar moment mode called '//trim(scalar_moment_mode))
        endif
    elseif (mtutil_mode.eq.'mij2pnt') then
        call mij2pnt(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values)
    elseif (mtutil_mode.eq.'mij2sdr') then
        nvals_output = 6
        call mij2sdr(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values(1),output_values(2),output_values(3),&
                     output_values(4),output_values(5),output_values(6))
    elseif (mtutil_mode.eq.'mij2sv') then
        nvals_output = 6
        call mij2sv(input_values(1),input_values(2),input_values(3),&
                    input_values(4),input_values(5),input_values(6),&
                    output_values(1:3),output_values(4:6))
    elseif (mtutil_mode.eq.'mij2ter') then
        call mij2ter(input_values(1),input_values(2),input_values(3),&
                     input_values(4),input_values(5),input_values(6),&
                     output_values(1),output_values(2),output_values(3))

    ! Input quantity: moment tensor eigenvectors/eigenvalues (P, N, T axes)
    elseif (mtutil_mode.eq.'pnt2dcp') then
        call pnt2dcp(input_values,output_values(1))
    elseif (mtutil_mode.eq.'pnt2mag') then
        if (scalar_moment_mode.eq.'dc') then
            call pnt2mag(input_values,output_values(1))
        elseif (scalar_moment_mode.eq.'nondc') then
            call pnt2mag_nondc(input_values,output_values(1))
        else
            call usage('mtutil: no scalar moment mode called '//trim(scalar_moment_mode))
        endif
    elseif (mtutil_mode.eq.'pnt2mij') then
        call pnt2mij(input_values,&
                     output_values(1),output_values(2),output_values(3),&
                     output_values(4),output_values(5),output_values(6))
    elseif (mtutil_mode.eq.'pnt2mom') then
        if (scalar_moment_mode.eq.'dc') then
            call pnt2mom(input_values,output_values(1))
        elseif (scalar_moment_mode.eq.'nondc') then
            call pnt2mom_nondc(input_values,output_values(1))
        else
            call usage('mtutil: no scalar moment mode called '//trim(scalar_moment_mode))
        endif
    elseif (mtutil_mode.eq.'pnt2sdr') then
        nvals_output = 6
        call pnt2sdr(input_values,&
                     output_values(1),output_values(2),output_values(3),&
                     output_values(4),output_values(5),output_values(6))
    elseif (mtutil_mode.eq.'pnt2sv') then
        nvals_output = 6
        call pnt2sv(input_values,output_values(1:3),output_values(4:6))
    elseif (mtutil_mode.eq.'pnt2ter') then
        call pnt2ter(input_values,output_values(1),output_values(2),output_values(3))

    ! Input quantity: double couple strike, dip, rake
    elseif (mtutil_mode.eq.'sdr2mij') then
        call sdr2mij(input_values(1),input_values(2),input_values(3),&
                     output_values(1),output_values(2),output_values(3),&
                     output_values(4),output_values(5),output_values(6))
    elseif (mtutil_mode.eq.'sdr2pnt') then
        call sdr2pnt(input_values(1),input_values(2),input_values(3),output_values)
    elseif (mtutil_mode.eq.'sdr2sdr') then
        nvals_output = 3
        call sdr2sdr(input_values(1),input_values(2),input_values(3),&
                     output_values(1),output_values(2),output_values(3))
    elseif (mtutil_mode.eq.'sdr2sv') then
        nvals_output = 3
        call sdr2sv(input_values(1),input_values(2),input_values(3),&
                    output_values(1:3))
    elseif (mtutil_mode.eq.'sdr2ter') then
        call sdr2ter(input_values(1),input_values(2),input_values(3),&
                     output_values(1),output_values(2),output_values(3))


    ! output -> kagan angle
    elseif (mtutil_mode.eq.'sdr2kagan') then
        call sdr2kagan(input_values(1), input_values(2), input_values(3), &
                       input_values2(1),input_values2(2),input_values2(3), &
                       output_values(1))
    elseif (mtutil_mode.eq.'mij2kagan') then
        call mij2kagan(input_values(1), input_values(2), input_values(3), &
                       input_values(4), input_values(5), input_values(6), &
                       input_values2(1),input_values2(2),input_values2(3), &
                       input_values2(4),input_values2(5),input_values2(6), &
                       output_values(1))

    else
        call usage('mtutil: no option '//trim(mtutil_mode))
    endif

    ! Print output quantity
    j = index(mtutil_mode,'2')
    output_mode = mtutil_mode(j+1:len(mtutil_mode))
    if (output_mode.eq.'mag') then
        fmt_string = '(1F10.4)'
    elseif (output_mode.eq.'mom') then
        fmt_string = '(1P1E14.6)'
    elseif (output_mode.eq.'mij') then
        fmt_string = '(1P6E14.6)'
    elseif (output_mode.eq.'pnt') then
        fmt_string = '(3(3F12.8,X),X,1P3E14.6)'
    elseif (output_mode.eq.'sdr') then
        if (nvals_output.eq.3) then
            fmt_string = '(F12.6,F12.6,F12.4)'
        else
            fmt_string = '(F12.6,F12.6,F12.4,4X,F12.6,F12.6,F12.4)'
        endif
    elseif (output_mode.eq.'ter') then
        fmt_string = '(3F12.8)'
    elseif (output_mode.eq.'dcp') then
        fmt_string = '(1F12.8)'
    elseif (output_mode.eq.'sv') then
        if (nvals_output.eq.3) then
            fmt_string = '(3F12.8)'
        else
            fmt_string = '(3F12.8,2X,3F12.8)'
        endif
    else
        write(fmt_string,'("(1P",I3,"E14.6)")') nvals_output
    endif
    write(luout,fmt_string) output_values(1:nvals_output)

enddo
1001 continue


! Close the input file and delete it if necessary
if (input_mode.eq.'file') then
    close(luin)
elseif (input_mode.eq.'cmdln') then
    close(luin,status='delete')
endif

! Close the output file
if (output.ne.'stdout') then
    close(luout)
endif

end

!--------------------------------------------------------------------------------------------------!

subroutine gcmdln()

use io, only: isNumeric
use mtutil, only: mtutil_mode, &
                  input, &
                  input_mode, &
                  output, &
                  nvals_input, &
                  nvals_output, &
                  scalar_moment_mode
implicit none

! Local variables
integer :: i, j, narg
character(len=512) :: tag

! Initialize
mtutil_mode = ''
input = ''
input_mode = ''
output = 'stdout'
nvals_input = 0
nvals_output = 0
scalar_moment_mode = 'dc'

! Number of arguments
narg = command_argument_count()
if (narg.eq.0) then
  call usage('')
endif


! First argument defines input parameters
i = 1
call get_command_argument(i,tag)
if (trim(tag).eq.'-sdr') then
    mtutil_mode = 'sdr2'
    nvals_input = 3
elseif (trim(tag).eq.'-mij') then
    mtutil_mode = 'mij2'
    nvals_input = 6
elseif (trim(tag).eq.'-pnt') then
    mtutil_mode = 'pnt2'
    nvals_input = 12
elseif (trim(tag).eq.'-mag') then
    mtutil_mode = 'mag2'
    nvals_input = 1
elseif (trim(tag).eq.'-mom') then
    mtutil_mode = 'mom2'
    nvals_input = 1
else
    call usage('mtutil: no input option '//tag)
endif
! write(0,*) trim(mode), nvals_input


! Optional argument defines a file name or value(s) for input
i = i + 1
if (i.gt.narg) then
    call usage('mtutil: input and output are required')
endif
! Determine how to read the second argument
call get_command_argument(i,input)
j = index(input,',')
if (j.ne.0) then
    ! comma => list of numbers to be read directly
    input_mode = 'cmdln'
    i = i + 1
else
    ! no comma => single number OR file name OR blank
    if (isNumeric(input)) then
        ! argument is readable as floating point number => read directly
        input_mode = 'cmdln'
        i = i + 1
    else
        ! argument is not readable as floating point number
        j = index(input,'-')
        if (j.eq.1) then
            ! argument starts with dash => mtutil flag
            input_mode = 'stdin'
        else
            ! argument does not start with dash => file name
            input_mode = 'file'
            i = i + 1
        endif
    endif
endif
! Make sure some input mode is specified
if (input_mode.eq.'') then
    call usage('mtutil: no input mode specified')
endif
! write(0,*) trim(input_mode)


! Second argument defines output parameters
if (i.gt.narg) then
    call usage('mtutil: an output option is required')
endif
call get_command_argument(i,tag)
if (trim(tag).eq.'-sdr') then
    mtutil_mode = trim(mtutil_mode)//'sdr'
    nvals_output = 3
elseif (trim(tag).eq.'-mij') then
    mtutil_mode = trim(mtutil_mode)//'mij'
    nvals_output = 6
elseif (trim(tag).eq.'-pnt') then
    mtutil_mode = trim(mtutil_mode)//'pnt'
    nvals_output = 12
elseif (trim(tag).eq.'-mag') then
    mtutil_mode = trim(mtutil_mode)//'mag'
    nvals_output = 1
elseif (trim(tag).eq.'-mom') then
    mtutil_mode = trim(mtutil_mode)//'mom'
    nvals_output = 1
elseif (trim(tag).eq.'-dcp') then
    mtutil_mode = trim(mtutil_mode)//'dcp'
    nvals_output = 1
elseif (trim(tag).eq.'-ternary') then
    mtutil_mode = trim(mtutil_mode)//'ter'
    nvals_output = 3
elseif (trim(tag).eq.'-sv') then
    mtutil_mode = trim(mtutil_mode)//'sv'
    nvals_output = 6
elseif (trim(tag).eq.'-kagan') then
    mtutil_mode = trim(mtutil_mode)//'kagan'
    nvals_output = 1
    nvals_input = -nvals_input
else
      call usage('mtutil: no output option '//tag)
endif
! write(0,*) trim(mode)


! Optional argument defines a file name for output
i = i + 1
if (i.le.narg) then
    call get_command_argument(i,output)
    j = index(output,'-')
    if (j.eq.1) then
        ! argument starts with dash => mtutil flag, so send output to stdout
        output = 'stdout'
    else
        ! argument does not start with dash => output file name
        i = i + 1
    endif
endif


! Extra arguments
do while (i.le.narg)
    call get_command_argument(i,tag)
    if (trim(tag).eq.'-dc') then
        scalar_moment_mode = 'dc'
    elseif (trim(tag).eq.'-non-dc') then
        scalar_moment_mode = 'nondc'
    else
        call usage('mtutil: no option '//trim(tag))
    endif
    i = i + 1
enddo

return
end subroutine

!--------------------------------------------------------------------------------------------------!

subroutine usage(str)
use io, only: stderr
implicit none
character(len=*) :: str

if (str.ne.'') then
    write(stderr,*) trim(str)
    write(stderr,*)
endif

write(stderr,*) 'Usage: mtutil -opt [INPUT] -opt [FILE] [...OPTIONS...]'
write(stderr,*)
write(stderr,*) 'Input/output:'
write(stderr,*) '-sdr      Strike, dip, and rake angles'
write(stderr,*) '-mij      Moment tensor components (mrr mtt mpp mrt mrp mtp)'
write(stderr,*) '-pnt      P, N, T axes (px,py,pz,nx,ny,nz,tx,ty,tz,pmag,nmag,tmag)'
write(stderr,*) '-mag      Moment magnitude'
write(stderr,*) '-mom      Seismic moment (Nm)'
write(stderr,*)
write(stderr,*) 'Output only:'
write(stderr,*) '-ternary  Ternary classification (fth fss fno)'
write(stderr,*) '-dcp      Double couple percentage'
write(stderr,*) '-sv       Slip vector (e n z components)'
write(stderr,*) '-kagan    Calculate angle between two moment tensors (requires double inputs)'
write(stderr,*)
write(stderr,*) 'Other options:'
write(stderr,*) '-dc       Use the double couple scalar seismic moment for -mom/-mag (default)'
write(stderr,*) '-non-dc   Use the general scalar seismic moment for -mom/-mag'
write(stderr,*)
write(stderr,*) 'Input format options:'
write(stderr,*) 'file name: read from file INPUT'
write(stderr,*) 'comma delimited list: read from command line'
write(stderr,*) '"stdin" or blank: read from standard input'
write(stderr,*)
write(stderr,*) 'Output format options:'
write(stderr,*) 'file name: write to file OUTPUT'
write(stderr,*) 'blank: print to standard output'
write(stderr,*)

call error_exit(1)
end subroutine
